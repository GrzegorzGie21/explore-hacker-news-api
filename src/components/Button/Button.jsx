import PropTypes from 'prop-types';

const Button = ({children, type, className, onClick}) =>
  <button type={type} onClick={onClick} className={className}>{children}</button>;

Button.propTypes = {
  type: PropTypes.string,
  onClick: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
};
Button.defaultProps = {
  type: 'button',
  className: '',
};
export default Button;