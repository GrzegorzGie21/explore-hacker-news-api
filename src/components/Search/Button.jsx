const Button = ({children, type='button', onClick}) => <button type={type} onClick={onClick}>{children}</button>
export default Button;