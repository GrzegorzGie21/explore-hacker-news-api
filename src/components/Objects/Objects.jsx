import searchItemByTitle from '../../utils';
import Button from '../Search/Button';

const Objects = ({items, title, removeItem}) =>
  <div>
    {items.filter(searchItemByTitle(title)).map(item => (
        <div key={item.id}>
            <span>
              <a href={item.url}>{item.title}</a>
            </span>
          <span>{item.author}</span>
          <span>{item.num_comments}</span>
          <span>{item.points}</span>
          <span>
              <Button onClick={() => removeItem(item.id)}>Remove</Button>
            {/*<button type='button' onClick={() => removeItem(item.id)}>Remove</button>*/}
            </span>
        </div>
      ),
    )
    }
  </div>;


export default Objects;