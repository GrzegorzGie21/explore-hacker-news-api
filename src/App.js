import {useState, useEffect} from 'react';
import './App.css';

import SearchForm from './components/Search';
import Objects from './components/Objects';
import Button from './components/Button';

const DEFAULT_QUERY = 'redux';
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = 100;

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

const App = ({api}) => {
  const [result, setResult] = useState(null);
  const [searchTerm, setSearchTerm] = useState('');
  
  useEffect(() => {
      try {
        fetchData();
      } catch (e) {
        console.log(e);
      }
    }, [],
  );
  
  const fetchData = async (searchTerm = DEFAULT_QUERY, currPage = DEFAULT_PAGE, hpp = DEFAULT_HPP) => {
    let data = await fetch(`${PATH_BASE}/${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${currPage}&${PARAM_HPP}${hpp}`);
    data = await data.json();
    const {hits, page} = data;
    const oldHits = page === 0 ? [] : result.hits;
    const updatedHits = [...oldHits, ...hits];
    setResult({hits: updatedHits, page});
  };
  
  //this method is used when JSON Server was active
  const removeItemHandler = async (id) => {
    try {
      await fetch(`${api}/${id}`, {
        method: 'DELETE',
      });
    } catch (e) {
      console.log(e);
    } finally {
      const checkId = item => item.objectID !== id;
      setResult([...result.hits].filter(checkId));
    }
  };
  
  const dismissItemHandler = (id) => {
    const checkId = item => item.objectID !== id;
    const updatedHits = result.hits.filter(checkId);
    setResult({...result, hits: updatedHits});
  };
  
  const handleChange = (e) => {
    const title = e.target.value;
    setSearchTerm(title);
  };
  
  const searchTermHandler = (e) => {
    fetchData(searchTerm);
    e.preventDefault();
  };
  
  const nextPageHandler = (e) => {
    const currentPage = result.page || 0;
    fetchData(searchTerm || DEFAULT_QUERY, currentPage + 1);
    // e.preventDefault();
  };
  
  // if (!result) return null;
  return (
    <div className="page">
      <div className={'interactions'}>
        <SearchForm title={searchTerm}
                    handleInput={handleChange}
                    handleSubmit={searchTermHandler}>
          Search
        </SearchForm>
      </div>
      {result && <Objects items={result.hits} removeItem={dismissItemHandler}/>}
      <div className={'interactions'}>
        {result && <Button onClick={nextPageHandler}>Next</Button>}
      </div>
    </div>
  );
};
export default App;
