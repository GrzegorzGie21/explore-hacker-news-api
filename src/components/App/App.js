import {useEffect, useState, useRef} from 'react';
import './App.css';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {library} from '@fortawesome/fontawesome-svg-core';
import {fas} from '@fortawesome/free-solid-svg-icons';

import SearchForm from '../Search';
import Objects from '../Objects';
import Button from '../Button';
import Loading from '../Loading';

import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  DEFAULT_PAGE,
  
  PARAM_HPP,
  PARAM_PAGE,
  PARAM_SEARCH,
  PATH_SEARCH,
  PATH_BASE,
} from '../../constants';

library.add(fas);

const App = () => {
  const [results, setResults] = useState(null);
  const [searchTerm, setSearchTerm] = useState(DEFAULT_QUERY);
  const [searchKey, setSearchKey] = useState(searchTerm);
  const input = useRef(null);
  const [isLoading, setIsLoading] = useState(false);
  
  const fetchData = async (searchTerm, currPage = DEFAULT_PAGE) => {
    setIsLoading(true);
    const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${currPage}&${PARAM_HPP}${DEFAULT_HPP}`;
    let data = await fetch(url);
    data = await data.json();
    setSearchTopStories(data);
  };
  
  const setSearchTopStories = ({hits, page}) => {
    console.log(hits, page);
    const oldHits = results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];
    setResults({...results, [searchKey]: {hits: updatedHits, page}});
    setIsLoading(false);
  };
  useEffect(() => setSearchKey(searchTerm), [searchTerm]);
  // useEffect(() => setIsLoading(true), [results]);
  
  useEffect(() => {
      try {
        fetchData(searchTerm);
      } catch (e) {
        console.log(e);
      }
      // input.current.focus();
    }, [],
  );
  
  //this method is used when JSON Server was active
  const removeItemHandler = async (id) => {
    const api = 'http://localhost:3001/objects';
    try {
      await fetch(`${api}/${id}`, {
        method: 'DELETE',
      });
    } catch (e) {
      console.log(e);
    } finally {
      const checkId = item => item.objectID !== id;
      setResults([...results.hits].filter(checkId));
    }
  };
  
  const dismissItemHandler = (id) => {
    const checkId = item => item.objectID !== id;
    const page = results[searchKey].page;
    const updatedHits = results[searchKey].hits.filter(checkId);
    setResults({...results, [searchKey]: {hits: updatedHits, page}});
  };
  
  const handleChange = (e) => {
    const title = e.target.value;
    setSearchTerm(title);
  };
  
  const shouldSearchApi = searchTerm => !results[searchTerm];
  
  const searchTermHandler = (e) => {
    console.log(searchKey);
    console.log(searchTerm);
    if (shouldSearchApi(searchTerm)) fetchData(searchTerm);
    e.preventDefault();
  };
  
  const nextPageHandler = (e) => {
    const currentPage = (results && results[searchKey] && results[searchKey].page) || 0;
    fetchData(searchTerm, currentPage + 1);
  };
  
  const searchClickHandler = () => {
    input.current.focus();
  };
  
  // if (isLoading) return <Loading/>;
  return (
    <div className="page">
      <div className={'interactions'}>
        <SearchForm title={searchTerm}
                    handleInput={handleChange}
                    handleSubmit={searchTermHandler}
                    input={input}
                    onClick={searchClickHandler}>
          Search
        </SearchForm>
      </div>
      {results && results[searchKey] && <Objects items={results[searchKey].hits} removeItem={dismissItemHandler}/>}
      <div className={'interactions'}>
        {isLoading ? <FontAwesomeIcon icon={'spinner'} size={'3x'} pulse/> : <Button onClick={nextPageHandler}>Next</Button>}
      </div>
    </div>
  );
};

export default App;
