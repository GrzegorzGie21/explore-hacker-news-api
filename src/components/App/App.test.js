import {render, screen} from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import App from './App';
import SearchForm from '../Search';
import Button from '../Button';
import Objects from '../Objects';

// test('renders learn react link', () => {
//   render(<App/>);
//   const linkElement = screen.getByText('Title');
//   expect(linkElement).toBeInTheDocument();
// });

describe('App', () => {
  it('renders App component', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App/>, div);
  });
  
  test('snapshots for App component', () => {
    const component = renderer.create(<App/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('SearchForm', () => {
  const props = {
    title: 'Search',
    onChange: () => {
    },
    handleInput: () => {
    },
    handleSubmit: () => {
    },
    children: 'Next',
  };
  
  it('renders SearchForm component', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SearchForm {...props}>Search</SearchForm>, div);
  });
  
  test('snapshots for SearchForm component', () => {
    const component = renderer.create(<SearchForm {...props}>Search</SearchForm>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  it('renders Button component', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>More</Button>, div);
  });
  
  test('snapshots for Button component', () => {
    const component = renderer.create(<Button>Next</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Objects', () => {
  const props = {
    items: [
      {title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
      {title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
    ],
    removeItem: () => {
    },
  };
  
  it('renders Objects component', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Objects {...props}/>, div);
  });
  
  test('snapshots for Object Component', () => {
    const component = renderer.create(<Objects {...props}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  
  // it('should show two items in list', function () {
  //   const element = shallow(<Objects {...props}/>);
  //   expect(element.find('.table-row').length).toBe(2)
  // });
});