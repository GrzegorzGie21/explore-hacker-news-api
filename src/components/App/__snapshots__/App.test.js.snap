// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots for App component 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="searchTitle"
      >
        <input
          name="searchTitle"
          onChange={[Function]}
          placeholder="redux"
          type="text"
          value="redux"
        />
        <button
          className=""
          type="submit"
        >
          Search
        </button>
      </label>
    </form>
  </div>
  <div
    className="interactions"
  />
</div>
`;

exports[`Button snapshots for Button component 1`] = `
<button
  className=""
  type="button"
>
  Next
</button>
`;

exports[`Objects snapshots for Object Component 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span>
      Title
    </span>
    <span>
      Author
    </span>
    <span>
      Number of comments
    </span>
    <span>
      Points
    </span>
    <span>
      Action
    </span>
  </div>
  <div
    className="table-row"
  >
    <span>
      <a>
        1
      </a>
    </span>
    <span>
      1
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Remove
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span>
      <a>
        2
      </a>
    </span>
    <span>
      2
    </span>
    <span>
      1
    </span>
    <span>
      2
    </span>
    <span>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Remove
      </button>
    </span>
  </div>
</div>
`;

exports[`SearchForm snapshots for SearchForm component 1`] = `
<form
  onSubmit={[Function]}
>
  <label
    htmlFor="searchTitle"
  >
    <input
      name="searchTitle"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="Search"
    />
    <button
      className=""
      type="submit"
    >
      Search
    </button>
  </label>
</form>
`;
